### PR TITLE
fixes #101: support new.target meta property

### LINKED ISSUE #101: ES6: new.target
# [12.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-left-hand-side-expressions) Left-Hand-Side Expressions
### Syntax

<pre>
<i>MemberExpression</i>[Yield] :
    <i>MetaProperty</i>

<i>MetaProperty</i> :
    <i>NewTarget</i>

<i>NewTarget</i> :
    <b>new</b> <b>.</b> <b>target</b>
</pre>

#### [15.1.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-scripts-static-semantics-early-errors)

<pre>
ScriptBody : StatementList
</pre>

- It is a Syntax Error if StatementList Contains NewTarget unless the source code containing NewTarget is eval code that is being processed by a direct eval that is contained in function code.
#### [15.2.1.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-module-semantics-static-semantics-early-errors)

<pre>
ModuleBody : ModuleItemList
</pre>

- It is a Syntax Error if ModuleItemList Contains NewTarget

### PR DESCRIPTION
**Pull request by michaelficarra**
Fixes #101.

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -1319,9 +1319,18 @@ export class Parser extends Tokenizer {
   parseNewExpression() {
     let startLocation = this.getLocation();
     this.expect(TokenType.NEW);
+    if (this.inFunctionBody && this.eat(TokenType.PERIOD)) {
+      let ident = this.expect(TokenType.IDENTIFIER);
+      if (ident.value !== "target" || ident.slice.end - ident.slice.start !== 6) {
```
Comment (ikarienator) [inline src/parser.js]: The size test is not necessary:

> All interpretations of IdentifierName within this specification are based upon their actual code points regardless of whether or not an escape sequence was used to contribute any particular code point.
> https://people.mozilla.org/~jorendorff/es6-draft.html#sec-names-and-keywords 

Exactly why we don't need to check the size of other contextual keywords.

File: test/other-expressions/new-target-expression.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        new Shift.ExpressionStatement(new Shift.NewExpression(new Shift.ArrayExpression([
+          new Shift.LiteralStringExpression("target"),
+        ]), [])),
+      ]))
+    );
+
+    testParseFailure("new.target;", "Unexpected token .");
+    testParseFailure("function f() { new.anythingElse; }", "Unexpected identifier");
+    testParseFailure("function f() { new..target; }", "Unexpected token .");
+    testParseFailure("function f() { new.\\u0074arget; }", "Unexpected identifier");
```
Comment (ikarienator) [inline test/other-expressions/new-target-expression.js]: This should be a success test.

**Comment by ikarienator** [comment]
Please rebase.

**Comment by michaelficarra** [comment]
Ping @ikarienator.

**Pull request merged**